src/sinks/aws_cloudwatch_metrics.rs

```diff
@@ -0,0 +1,381 @@
+    ) -> Result<super::RouterSink, String> {
+        let client = Self::create_client(config.region.clone().try_into()?)?;
+
+        let batch_size = config.batch_size.unwrap_or(5);
+        let batch_timeout = config.batch_timeout.unwrap_or(1);
+
+        let timeout = config.request_timeout_secs.unwrap_or(30);
+        let in_flight_limit = config.request_in_flight_limit.unwrap_or(5);
+        let rate_limit_duration = config.request_rate_limit_duration_secs.unwrap_or(1);
+        let rate_limit_num = config.request_rate_limit_num.unwrap_or(5);
+        let retry_attempts = config.request_retry_attempts.unwrap_or(usize::max_value());
+        let retry_backoff_secs = config.request_retry_backoff_secs.unwrap_or(1);
+
+        let policy = FixedRetryPolicy::new(
+            retry_attempts,
+            Duration::from_secs(retry_backoff_secs),
+            CloudWatchMetricsRetryLogic,
+        );
+
+        let cloudwatch_metrics = CloudWatchMetricsService { client, config };
+
+        let svc = ServiceBuilder::new()
+            .concurrency_limit(in_flight_limit)
+            .rate_limit(rate_limit_num, Duration::from_secs(rate_limit_duration))
+            .retry(policy)
+            .timeout(Duration::from_secs(timeout))
+            .service(cloudwatch_metrics);
+
+        let sink = BatchServiceSink::new(svc, acker).partitioned_batched_with_min(
+            Vec::new(),
+            batch_size,
+            Duration::from_secs(batch_timeout),
+        );
+
+        Ok(Box::new(sink))
+    }
+
+    fn healthcheck(config: &CloudWatchMetricsSinkConfig) -> Result<super::Healthcheck, String> {
+        let client = Self::create_client(config.region.clone().try_into()?)?;
+
+        let datum = MetricDatum {
+            metric_name: "healthcheck".into(),
+            value: Some(1.0),
+            ..Default::default()
+        };
+        let request = PutMetricDataInput {
+            namespace: config.namespace.clone(),
+            metric_data: vec![datum],
+        };
+
+        let response = client.put_metric_data(request);
+        let healthcheck = response.map_err(|err| err.to_string());
+
+        Ok(Box::new(healthcheck))
+    }
+
+    fn create_client(region: Region) -> Result<CloudWatchClient, String> {
+        #[cfg(test)]
+        {
+            // Moto (used for mocking AWS) doesn't recognize 'custom' as valid region name
+            let region = match region {
+                Region::Custom { endpoint, .. } => Region::Custom {
+                    name: "us-east-1".into(),
+                    endpoint,
+                },
+                _ => panic!("Only Custom regions are supported for CloudWatchClient testing"),
+            };
+            Ok(CloudWatchClient::new(region))
+        }
+
+        #[cfg(not(test))]
+        {
+            Ok(CloudWatchClient::new(region))
+        }
+    }
+}
+
+use bytes::Bytes;
+
+impl Partition<Bytes> for Event {
+    fn partition(&self) -> Bytes {
+        "key".into()
```

Comment:
I was thinking about partitioning for some time, and I didn't come with any better scheme to use. Maybe we should open a follow up issue to investigate further payload structure optimisations.

Namespaces are currently configured at Sink level (not sure if that's what users will find useful /cc @binarylogic), thus they will not work as partition keys.

Partitioning by metric name will not bring much value, because:
a) in some cases we will need to pass timestamps, and timestamp is part of `MetricDatum`, and timestamped datums cannot really be compressed
b) labels aka 'dimensions' are considered part of metric name, and those could have some extra variability, and also are part of `MetricDatum`.

To summarise, seems like the only simple batching strategy we can implement is to pack 20 `MetricDatum` objects into `PutMetricData` request, and in each `MetricDatum` we will not benefit from any kind of partitioning. Hope this all makes sense!

Reply:
I think maybe partitioning shouldn't be in the scope of this PR then. I think we could use some more discussion around that. In that case, this PR should be updated to reflect that. That means instead the sink combinator should be changed to just batched_with_min.

Reply:
Omg how I just overcomplicated it :)